aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -267,35 +453,24 @@ Tensor batch_norm_backward_elemt_cuda(const Tensor& self, const Tensor& input, c
 }
 
 std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda(
-        const Tensor& self, const c10::optional<Tensor>& running_mean_opt, const c10::optional<Tensor>& running_var_opt, double momentum) {
-  // See [Note: hacky wrapper removal for optional tensor]
-  c10::MaybeOwned<Tensor> running_mean_maybe_owned = at::borrow_from_optional_tensor(running_mean_opt);
-  const Tensor& running_mean = *running_mean_maybe_owned;
-  const Tensor& running_var = c10::value_or_else(running_var_opt, [] {return Tensor();});
+    const Tensor& self, const c10::optional<Tensor>& running_mean_opt,
+    const c10::optional<Tensor>& running_var_opt, double momentum) {
+  c10::MaybeOwned<Tensor> running_mean = at::borrow_from_optional_tensor(running_mean_opt);
+  c10::MaybeOwned<Tensor> running_var = at::borrow_from_optional_tensor(running_var_opt);
 
-  return AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, self.scalar_type(), "batch_norm_backward", [&] {
-    auto mean_st = running_mean.dtype();
-    auto var_st = running_var.dtype();
-    TORCH_CHECK(mean_st == var_st, "running_mean and running_var need to have the same data types");
-    // <sigh> Some workloads depend on passing in half input and float stats, which is
-    // usually handled by cuDNN. However, the JIT sometimes replaces cuDNN calls with this
-    // one so it needs to support the same case, or people start to complain.
-    bool is_half_float = std::is_same<scalar_t, at::Half>::value && mean_st == at::kFloat;
-    bool is_bfloat16_float = std::is_same<scalar_t, at::BFloat16>::value && mean_st == at::kFloat;
-    if (cuda::detail::canUse32BitIndexMath(self)) {
-      if (is_half_float || is_bfloat16_float) {
-        return batch_norm_update_stats_cuda_template<scalar_t, float, int32_t>(self, running_mean, running_var, momentum);
-      } else {
-        return batch_norm_update_stats_cuda_template<scalar_t, scalar_t, int32_t>(self, running_mean, running_var, momentum);
-      }
-    } else {
-      if (is_half_float || is_bfloat16_float) {
-        return batch_norm_update_stats_cuda_template<scalar_t, float, int64_t>(self, running_mean, running_var, momentum);
-      } else {
-        return batch_norm_update_stats_cuda_template<scalar_t, scalar_t, int64_t>(self, running_mean, running_var, momentum);
-      }
-    }
-  });
+  const int64_t n_input = self.size(1);
+  auto options = self.options().dtype(
+      at::toAccumulateType(self.scalar_type(), /*is_cuda=*/true));
+  auto save_mean = at::empty({n_input}, options);
+  auto save_var = at::empty({n_input}, options);
+
+  batch_norm_mean_var(self, save_mean, save_var);
+  TORCH_CHECK(running_mean->defined() == running_var->defined());
+  if (running_mean->defined()) {
+    const int64_t N = self.numel() / save_mean.numel();
+    batch_norm_update_stats(save_mean, save_var, *running_mean, *running_var, momentum, N);
```

Comment:
`scalar_t` in that function means the dtype of `running_mean`/`running_var` which can be low precision,  so there is no issue.

Reply:
Actually, maybe it isn't working out like that because I dispatch on `common_type`. Changed to explicitly take scalar_type from `running_mean`/`running_var`.

Reply:
`batch_norm_update_stats` still dispatches on `common_type`, so if `running_mean` is half, `scalar_t` will still be float, because `common_dtype` is float, and dtypes in lambda would be wrong